cpp/src/Slice/Grammar.y

```diff
@@ -1765,112 +1782,81 @@ enumerator_initializer
 ;
 
 // ----------------------------------------------------------------------
-out_qualifier
+parameter
 // ----------------------------------------------------------------------
-: ICE_OUT
-{
-    auto out = make_shared<BoolTok>();
-    out->v = true;
-    $$ = out;
-}
-| %empty
+: optional_type_id
 {
-    auto out = make_shared<BoolTok>();
-    out->v = false;
-    $$ = out;
-}
-;
+    auto tsp = dynamic_pointer_cast<OptionalDefTok>($1);
+    ParameterPtr param;
 
-// ----------------------------------------------------------------------
-parameters
-// ----------------------------------------------------------------------
-: parameter_list
-| %empty
-;
-
-// ----------------------------------------------------------------------
-parameter_list
-// ----------------------------------------------------------------------
```

Comment:
I split `parameter_list` into 2 rules: `parameter` and `paramter_list`.
All the meat is in parameter, and now `parameter_list` is really just for parsing comma-separated parameters.
The current logic was too ugly, where we duplicate each possible match, one with a leading comma and one without,